libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -182,28 +174,31 @@ struct gic_dist_map {
     uint32_t component_id[4];                           /* [0xFF0, 0xFFF] */
 };
 
-#define MAX_LR_OVERFLOW 64
+/* TODO: A typical number of list registers supported by GIC is four, but not
+ * always. One particular way to probe the number of registers is to inject a
+ * dummy IRQ with seL4_ARM_VCPU_InjectIRQ(), using LR index high enough to be
+ * not supported by any target; the kernel will reply with the supported range
+ * of LR indexes.
+ */
+#define NUM_LIST_REGS 4
+/* This is a rather abritrary number, increase if needed. */
+#define MAX_IRQ_QUEUE_LEN 64
+#define IRQ_QUEUE_NEXT(_i) (((_i) + 1) & (MAX_IRQ_QUEUE_LEN - 1))
+
+static_assert((MAX_IRQ_QUEUE_LEN & (MAX_IRQ_QUEUE_LEN - 1)) == 0,
+              "IRQ ring buffer size must be power of two");
 
-struct lr_of {
-    struct virq_handle irqs[MAX_LR_OVERFLOW]; /* circular buffer */
+struct irq_queue {
+    struct virq_handle *irqs[MAX_IRQ_QUEUE_LEN]; /* circular buffer */
     size_t head;
     size_t tail;
-    bool full;
 };
 
-static inline size_t LR_OF_NEXT(size_t idx)
```

Comment:
Please leave this in unless there is a strong reason to go back to the macro.

Reply:
I can do that, but I would like you to first review my patch set fully; I'm getting rid of that `full` field of `struct lr_of` and ring buffer code is much more comprehensible now IMHO. I was thinking my `static_assert `about the requirement of size of power of two and the and-operation for finding out the next index would be self-documenting. TBH I found your addition not exactly helping understanding the code and feel that the data structure is implemented in non-optimal way if it needs helpers/comments like that. But not really wanting to start a flame war, I just would you to read all my code first.

Reply:
Sorry, my bad. Looked like a merge conflict to me first, but with the 2^n size this is better. 
My argument for the function is, that this allows a strong binding to the actual array size in the struct and not just a constant. It's supposed to make the code more robust and the compiler is able to optimize this like the macro.

Reply:
But that stronger array binding could also be done in the macro. I may make a separate PR for this, so it's no blocker for this PR.

Reply:
Yeah, my intention was to ping you and ask how do you like my proposal, but you were faster. And I can see where you are coming from. We could have sanity checks with this 2^n also, but I guess then it boils down to more fundamental questions like a) what else should we assert and b) shouldn't we be doing the assert before using the index (right before dereferencing an array), not just when derive a new index.

Ideally the data structures should be in a separate header/source file, having them in between the code that uses them makes it quite hard to grasp sometimes. But something like that is a PITA to do in C.

Reply:
Forgot to mention that I am not that worried about performance in this context. It is just readability that matters to me.

Reply:
> My argument for the function is, that this allows a strong binding to the actual array size in the struct and not just a constant. It's supposed to make the code more robust and the compiler is able to optimize this like the macro.

I think this issue could be resolved outside of this PR.  The macro and array definitions are defined within 10 lines of each other and use the same size constant which is already good enough.